util/compression.h

```diff
@@ -1482,6 +1482,35 @@ inline std::string ZSTD_TrainDictionary(const std::string& samples,
 #endif  // ZSTD_VERSION_NUMBER >= 10103
 }
 
+inline std::string ZSTD_FinalizeDictionary(
+    const std::string& samples, const std::vector<size_t>& sample_lens,
+    size_t max_dict_bytes) {
+  // ZDICT_finalizeDictionary is only stable since version 1.4.5
+#if ZSTD_VERSION_NUMBER >= 10405  // v1.4.5+
+  assert(samples.empty() == sample_lens.empty());
+  if (samples.empty()) {
+    return "";
+  }
+  std::string dict_data(max_dict_bytes, '\0');
+  size_t dict_len = ZDICT_finalizeDictionary(
+      dict_data.data(), max_dict_bytes, samples.data(),
+      static_cast<size_t>(samples.size()), samples.data(), sample_lens.data(),
+      static_cast<unsigned>(sample_lens.size()), {});
+  if (ZDICT_isError(dict_len)) {
+    // fall back to use concatenated samples as raw content dictionary
+    return samples;
```

Comment:
This might cause the test failure - if the dictionary includes all data blocks, we'd get files of 2x target size (dictionary consumes 1x, actual data is the other 1x), and there'd be fewer files because compression ratio is better.

Reply:
Good catch, thanks!